darts/timeseries.py

```diff
@@ -4233,66 +4269,138 @@ def _fill_missing_dates(
+
+        if isinstance(time_index, pd.DatetimeIndex):
+            has_datetime_index = True
+            observed_freqs = cls._observed_freq_datetime_index(time_index)
+        else:  # integer index (non RangeIndex)
+            has_datetime_index = False
+            observed_freqs = cls._observed_freq_integer_index(time_index)
+
+        offset_alias_info = (
+            (
+                " For more information about frequency aliases, read "
+                "https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#offset-aliases"
+            )
+            if has_datetime_index
+            else ""
+        )
+        if not len(observed_freqs) == 1:
+            raise_log(
+                ValueError(
+                    f"Could not observe an inferred frequency. An explicit frequency must be evident over a span of "
+                    f"at least 3 consecutive time stamps in the input data. {offset_alias_info}"
+                    if not len(observed_freqs)
+                    else f"Could not find a unique inferred frequency (not constant). Observed frequencies: "
+                    f"{observed_freqs}. If any of those is the actual frequency, try passing it with "
+                    f"`fill_missing_dates=True` and `freq=your_frequency`.{offset_alias_info}"
+                ),
+                logger,
+            )
+
+        freq = observed_freqs.pop()
+
+        return cls._restore_xarray_from_frequency(sorted_xa, freq)
+
+    @staticmethod
+    def _sort_index(xa: xr.DataArray) -> xr.DataArray:
+        """Sorts an xarray by its time dimension index (only if it is not already monotonically increasing)."""
+        time_dim = xa.dims[0]
+        return (
             xa.copy()
             if xa.get_index(time_dim).is_monotonic_increasing
             else xa.sortby(time_dim)
         )
-        time_index = sorted_xa.get_index(time_dim)
-
-        offset_alias_info = (
-            "For more information about frequency aliases, read "
-            "https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#offset-aliases"
-        )
 
+    @staticmethod
+    def _observed_freq_datetime_index(index: pd.DatetimeIndex) -> set:
+        """Returns all observed/inferred frequencies of a pandas DatetimeIndex. The frequencies are inferred from all
+        combinations of three adjacent time steps
+        """
         step_size = 3
-        n_dates = len(time_index)
-        # this creates n steps containing 3 timestamps each; used to infer frequency of time_index
+        n_dates = len(index)
+        # this creates n steps containing 3 timestamps each; used to infer frequency of index
         steps = np.column_stack(
-            [time_index[i : (n_dates - step_size + (i + 1))] for i in range(step_size)]
+            [index[i : (n_dates - step_size + (i + 1))] for i in range(step_size)]
         )
         observed_freqs = {pd.infer_freq(step) for step in steps}
```

Comment:
Yes that's true. Maybe we could look unique time deltas only between time stamps `i+1` and `i`.
- find unique time deltas (similar to step size for integer index)
- infer the frequency for all these unique deltas (maybe twice per unique delta, once using (i-1, i, i+1) and the other with (i, i+1, i+2)
- continue as before

This will probably in most cases drastically reduce the examples `steps` we have to infer.
What do you think @hrzn ?

Reply:
@hrzn. Okay, so I came up with an optimized approach here. This reduces time series creation time drastically. See below (left image shows creation time in 10^(n) seconds, right image shows time in seconds).
This new approach is around 300 times faster for larger series! (I didn't compute series length 10^7 for the old approach as it's too slow).

<img width="829" alt="image" src="https://user-images.githubusercontent.com/84317528/221407316-22852db8-a0c6-4143-ac0d-cff3f7350c91.png">

Reply:
Wow, that's super cool 👍 well done!